File: api/ingest.py
from http.server import BaseHTTPRequestHandler
import json, os, io, httpx
from pypdf import PdfReader
from supabase import create_client

SUPABASE_URL = os.environ["SUPABASE_URL"]
SUPABASE_KEY = os.environ["SUPABASE_SERVICE_ROLE_KEY"]
VOYAGE_API_KEY = os.environ["VOYAGE_API_KEY"]
BLOB_READ_WRITE_TOKEN = os.environ.get("BLOB_READ_WRITE_TOKEN", "")

CHUNK_TOKENS = 500
OVERLAP_TOKENS = 50


def rough_token_count(text: str) -> int:
    return max(1, len(text) // 4)


def chunk_text(pages: list[tuple[int, str]]) -> list[dict]:
    chunks: list[dict] = []
    buffer = ""
    buffer_page = 1

    for page_num, text in pages:
        paragraphs = [p.strip() for p in text.split("\n\n") if p.strip()]
        for para in paragraphs:
            if rough_token_count(buffer) + rough_token_count(para) > CHUNK_TOKENS and buffer:
                chunks.append({"content": buffer.strip(), "page_number": buffer_page})
                words = buffer.split()
                overlap_words = max(1, OVERLAP_TOKENS * 4 // 5)
                buffer = " ".join(words[-overlap_words:]) + " " if words else ""
                buffer_page = page_num
            buffer += para + "\n\n"
            if not buffer_page:
                buffer_page = page_num

    if buffer.strip():
        chunks.append({"content": buffer.strip(), "page_number": buffer_page})

    return chunks


def embed_batch(texts: list[str]) -> list[list[float]]:
    resp = httpx.post(
        "https://api.voyageai.com/v1/embeddings",
        headers={"Authorization": f"Bearer {VOYAGE_API_KEY}", "Content-Type": "application/json"},
        json={"input": texts, "model": "voyage-3", "input_type": "document"},
        timeout=60.0,
    )
    resp.raise_for_status()
    return [d["embedding"] for d in resp.json()["data"]]


def process_document(blob_url: str, filename: str, document_id: str, session_id: str) -> None:
    supabase = create_client(SUPABASE_URL, SUPABASE_KEY)

    headers = {"Authorization": f"Bearer {BLOB_READ_WRITE_TOKEN}"} if BLOB_READ_WRITE_TOKEN else {}
    resp = httpx.get(blob_url, headers=headers, follow_redirects=True, timeout=60.0)
    resp.raise_for_status()

    reader = PdfReader(io.BytesIO(resp.content))
    pages = [(i + 1, page.extract_text() or "") for i, page in enumerate(reader.pages)]

    chunks = chunk_text(pages)
    if not chunks:
        supabase.table("documents").update({"status": "error"}).eq("id", document_id).execute()
        return

    texts = [c["content"] for c in chunks]
    all_embeddings: list[list[float]] = []
    for i in range(0, len(texts), 100):
        all_embeddings.extend(embed_batch(texts[i : i + 100]))

    rows = [
        {
            "document_id": document_id,
            "session_id": session_id,
            "content": chunk["content"],
            "embedding": embedding,
            "chunk_index": idx,
            "page_number": chunk["page_number"],
        }
        for idx, (chunk, embedding) in enumerate(zip(chunks, all_embeddings))
    ]
    supabase.table("chunks").insert(rows).execute()
    supabase.table("documents").update({"status": "ready", "chunk_count": len(rows)}).eq("id", document_id).execute()


class handler(BaseHTTPRequestHandler):
    def do_POST(self):
        body: dict = {}
        try:
            length = int(self.headers.get("Content-Length", 0))
            body = json.loads(self.rfile.read(length))
            process_document(body["blob_url"], body["filename"], body["document_id"], body["session_id"])
            self._send(200, {"status": "ok"})
        except Exception as e:
            error_msg = str(e)
            # Mark document as errored so callers don't wait forever
            if body.get("document_id"):
                try:
                    supabase = create_client(SUPABASE_URL, SUPABASE_KEY)
                    supabase.table("documents").update({"status": "error"}).eq("id", body["document_id"]).execute()
                except Exception:
                    pass
            self._send(500, {"error": error_msg})

    def _send(self, code: int, data: dict) -> None:
        self.send_response(code)
        self.send_header("Content-Type", "application/json")
        self.end_headers()
        self.wfile.write(json.dumps(data).encode())

    def log_message(self, format, *args):
        pass

File: src/lib/ingest.ts
import { createServerClient } from '@/lib/supabase'
import { get } from '@vercel/blob'
import { inflateSync } from 'zlib'

const CHUNK_TOKENS = 500
const OVERLAP_TOKENS = 50
const VOYAGE_API_KEY = process.env.VOYAGE_API_KEY!

/**
 * Extract text from a PDF buffer.
 * Handles both plain and FlateDecode (zlib) compressed content streams.
 * Works in any Node.js environment without browser API dependencies.
 */
function extractPdfText(buffer: Buffer): string {
  const raw = buffer.toString('latin1')
  const texts: string[] = []

  const streamRegex = /stream\r?\n([\s\S]*?)\r?\nendstream/g
  let m: RegExpExecArray | null

  while ((m = streamRegex.exec(raw)) !== null) {
    const streamStart = m.index
    const streamData = m[1]

    // Check the preceding object dict for compression filter
    const preceding = raw.slice(Math.max(0, streamStart - 500), streamStart)
    const isFlate = preceding.includes('FlateDecode')

    let content: string
    if (isFlate) {
      try {
        const compressed = Buffer.from(streamData, 'latin1')
        const decompressed = inflateSync(compressed)
        content = decompressed.toString('utf8')
      } catch {
        continue // skip unreadable compressed streams
      }
    } else {
      content = streamData
    }

    if (!content.includes('Tj') && !content.includes('TJ')) continue

    // Extract from Tj operator: (text) Tj
    for (const t of content.matchAll(/\(([^)\\]*(?:\\.[^)\\]*)*)\)\s*Tj/g)) {
      const s = t[1]
        .replace(/\\n/g, '\n').replace(/\\r/g, '\r').replace(/\\t/g, '\t')
        .replace(/\\\(/g, '(').replace(/\\\)/g, ')').replace(/\\\\/g, '\\')
        .replace(/\\(\d{3})/g, (_, oct) => String.fromCharCode(parseInt(oct, 8)))
      if (s.trim()) texts.push(s)
    }

    // Extract from TJ operator: [(text) offset ...] TJ
    for (const t of content.matchAll(/\[([^\]]+)\]\s*TJ/g)) {
      for (const p of t[1].matchAll(/\(([^)\\]*(?:\\.[^)\\]*)*)\)/g)) {
        const s = p[1].replace(/\\n/g, '\n')
        if (s.trim()) texts.push(s)
      }
    }
  }

  return texts.join(' ').replace(/\s+/g, ' ').trim()
}

function roughTokenCount(text: string): number {
  return Math.max(1, Math.floor(text.length / 4))
}

function chunkText(pages: { pageNum: number; text: string }[]): { content: string; pageNumber: number }[] {
  const chunks: { content: string; pageNumber: number }[] = []
  let buffer = ''
  let bufferPage = 1

  for (const { pageNum, text } of pages) {
    const paragraphs = text.split(/\n\n+/).map(p => p.trim()).filter(Boolean)
    for (const para of paragraphs) {
      if (roughTokenCount(buffer) + roughTokenCount(para) > CHUNK_TOKENS && buffer) {
        chunks.push({ content: buffer.trim(), pageNumber: bufferPage })
        const words = buffer.split(/\s+/)
        const overlapWords = Math.max(1, Math.floor((OVERLAP_TOKENS * 4) / 5))
        buffer = words.slice(-overlapWords).join(' ') + ' '
        bufferPage = pageNum
      }
      buffer += para + '\n\n'
      if (!bufferPage) bufferPage = pageNum
    }
  }

  if (buffer.trim()) {
    chunks.push({ content: buffer.trim(), pageNumber: bufferPage })
  }

  return chunks
}

async function embedBatch(texts: string[], attempt = 0): Promise<number[][]> {
  const res = await fetch('https://api.voyageai.com/v1/embeddings', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${VOYAGE_API_KEY}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ input: texts, model: 'voyage-3', input_type: 'document' }),
  })

  if (res.status === 429 && attempt < 4) {
    const retryAfter = Number(res.headers.get('retry-after') ?? 0)
    const waitMs = retryAfter > 0 ? retryAfter * 1000 : Math.min(2 ** attempt * 5000, 60000)
    await new Promise(r => setTimeout(r, waitMs))
    return embedBatch(texts, attempt + 1)
  }

  if (!res.ok) throw new Error(`Voyage AI error: ${res.status} ${await res.text()}`)
  const data = await res.json()
  return data.data.map((d: { embedding: number[] }) => d.embedding)
}

export async function processDocument(
  blobUrl: string,
  filename: string,
  documentId: string,
  sessionId: string,
): Promise<void> {
  const supabase = createServerClient()

  try {
    // Download PDF from Vercel Blob using the SDK (required for private blobs)
    const blobResult = await get(blobUrl, { access: 'private' })
    if (!blobResult) throw new Error(`Blob not found: ${blobUrl}`)
    const pdfBuffer = Buffer.from(await new Response(blobResult.stream).arrayBuffer())
    const fullText = extractPdfText(pdfBuffer)

    if (!fullText || fullText.length < 10) {
      await supabase.from('documents').update({ status: 'error' }).eq('id', documentId)
      throw new Error(`No readable text found in "${filename}". The PDF may be image-based or encrypted.`)
    }

    const pages = [{ pageNum: 1, text: fullText }]
    const chunks = chunkText(pages)

    if (chunks.length === 0) {
      await supabase.from('documents').update({ status: 'error' }).eq('id', documentId)
      throw new Error(`Could not split "${filename}" into chunks.`)
    }

    // Embed in batches of 100
    const texts = chunks.map(c => c.content)
    const allEmbeddings: number[][] = []
    for (let i = 0; i < texts.length; i += 100) {
      const batch = await embedBatch(texts.slice(i, i + 100))
      allEmbeddings.push(...batch)
    }

    // Insert chunks into Supabase
    const rows = chunks.map((chunk, idx) => ({
      document_id: documentId,
      session_id: sessionId,
      content: chunk.content,
      embedding: allEmbeddings[idx],
      chunk_index: idx,
      page_number: chunk.pageNumber,
    }))

    await supabase.from('chunks').insert(rows)
    await supabase
      .from('documents')
      .update({ status: 'ready', chunk_count: rows.length })
      .eq('id', documentId)

    console.log(`[ingest] ${filename}: ${chunks.length} chunks ingested`)
  } catch (err) {
    console.error('[ingest] error:', err)
    await supabase.from('documents').update({ status: 'error' }).eq('id', documentId)
    throw err
  }
}
